src/core/ComboBox/ComboBox.tsx

```diff
@@ -117,252 +123,317 @@ export const ComboBox = <T,>(props: ComboBoxProps<T>) => {
 
   useTheme();
 
-  /** Generates a memoized id for an option, given the index from original list */
-  const getOptionId = React.useCallback(
-    (index: number) =>
-      options[index].id ??
-      `${id}-option${options.findIndex(
-        ({ value }) => value === options[index].value,
-      )}`,
-    [options, id],
-  );
-
-  const userOnChange = React.useRef(onChange);
-
-  const memoizedItems = React.useMemo(
-    () =>
-      options.map((option, index) => {
-        const { label, value, ...rest } = option;
-        const additionalProps = {
-          value: value,
-          role: 'option',
-          onClick: () => {
-            setSelectedValue(value);
-            userOnChange.current?.(value);
-            setIsOpen(false);
-          },
-        };
-        if (itemRenderer) {
-          return React.cloneElement(
-            itemRenderer(option, {
-              id: getOptionId(index),
-              index,
-              isSelected: false,
-              isFocused: false,
-            }),
-            additionalProps,
-          );
-        }
-        return (
-          <MenuItem
-            id={getOptionId(index)}
-            key={getOptionId(index)}
-            {...additionalProps}
-            {...rest}
-          >
-            {label}
-          </MenuItem>
-        );
-      }),
-    [options, getOptionId, itemRenderer],
-  );
+  // Add id and originalIndex to each option
+  const [options, setOptions] = React.useState<ComboBoxOption<T>[]>([]);
+  React.useLayoutEffect(() => {
+    setOptions(
+      optionsProp.map((option, index) => ({
+        ...option,
+        id: `${id}-option${index}`,
+        __originalIndex: index,
+      })),
+    );
+  }, [id, optionsProp]);
 
   const inputRef = React.useRef<HTMLInputElement>(null);
   const menuRef = React.useRef<HTMLUListElement>(null);
   const toggleButtonRef = React.useRef<HTMLSpanElement>(null);
-  const [isOpen, setIsOpen] = React.useState(false);
+  const optionsRef = React.useRef<Array<HTMLLIElement>>([]);
+
+  // Latest value of the onChange prop
+  const onChangeProp = React.useRef(onChange);
+  React.useEffect(() => {
+    onChangeProp.current = onChange;
```

Comment:
whats the reason to save it in ref, instead of using directly?

Reply:
i think this part is the same as before, it's to avoid requiring user to memoize onChange